pipeline/batch_pipeline.py

```diff
@@ -17,45 +18,70 @@
     if max_page_num == 0:
-        return "No new data to insert, exiting"
-    for page_num in track(
-        range(1, max_page_num + 1), description="Processing batch data..."
-    ):
-        gpt_response = []
-        data = get_listing_data(BATCH_URL, page_num)
-        count += len(data)
-
-        for index, item in enumerate(data):
-            court_case_citation = item.get("citation")
-
-            if r.exists(f"court_case:{court_case_citation}"):
-                contents = r.hgetall(f"court-case:{court_case_citation}")
-                case_details = json.loads(contents["case_details"])
-                gpt_response.append(case_details)
-                print(f"Cache hit for {court_case_citation}, retrieved from Redis.")
-            else:
-                response = get_data(data, index)
-                r.hset(
-                    f"court_case:{court_case_citation}",
-                    mapping={"case_details": json.dumps(response)},
-                )
-                gpt_response.append(response)
-                print(
-                    f"Cache miss for {court_case_citation}, GPT data fetched and stored in Redis."
-                )
-
-        table_data = assemble_data(gpt_response)
-        conn = get_connection()
-        insert_to_database(conn, table_data)
-        print(f"Page {page_num} inserted to database, {count} records in total")
-
-    return "Data inserted to database"
+        message = "No new data to insert, exiting"
+        logger.info(message)
+        return None
+
+    count = 0
+    with Progress() as progress:
+        task = progress.add_task("[cyan]Processing batch data...", total=max_page_num)
+        for page_num in range(1, max_page_num + 1):
+            data = get_listing_data(BATCH_URL, page_num)
+            count += len(data)
+            gpt_response = []
+
+            for index, item in enumerate(data):
+                court_case_citation = item.get("citation")
+                if r.exists(court_case_citation):
+                    case_details = literal_eval(
+                        r.hgetall(court_case_citation)["case_details"]
+                    )
+                    gpt_response.append(case_details)
+                    message = (
+                        f"Cache hit for {court_case_citation}, retrieved from Redis."
+                    )
+                else:
+                    response = get_data(data, index)
+                    r.hset(
+                        court_case_citation,
+                        mapping={"case_details": str(response)},
+                    )
+                    gpt_response.append(response)
+                    message = (
+                        f"Cache miss for {court_case_citation}, "
+                        "GPT data fetched and stored in Redis."
+                    )
+                logger.info(message)
+
+            insert_to_database(get_connection(), assemble_data(gpt_response, True))
```

Comment:
Is this main function not insanely large, perhaps break it down into smaller functions that complete the different parts such as one part that processes the website pages, another that handles the cache and a final function that inserts it?

Reply:
Yes, this is a good point, the main point for keeping it together at the moment is my understanding is that you have to keep passing the logger around as a function argument in the subsequent functions - an issue I encountered when doing this for the live_pipeline, definitely something we should double back around too if we have time.